File: vehicle/haskell/src/Aircraft.hs
module Aircraft (Aircraft (..)) where

import Vehicle as V (Vehicle)

class Vehicle aircraftType => Aircraft aircraftType where
  getEngine :: aircraftType -> String
  getEngine _ = "turbofan aircraft engines"

  pilot :: aircraftType -> IO ()
  pilot aircraft = do
    putStrLn ("Aircraft engine: " <> getEngine aircraft)
    putStrLn ">>>>>>>>>>>>>>>>>>>>>>>>>>>"

  simulate :: aircraftType -> IO ()
  simulate = pilot

File: vehicle/rust/src/aircraft.rs
use crate::Vehicle;

pub trait Aircraft: Vehicle {
  fn get_engine(&self) -> String {
    String::from("turbofan aircraft engines")
  }

  fn pilot(&self) {
    println!("Aircraft engine: {}", Aircraft::get_engine(self));
    println!(">>>>>>>>>>>>>>>>>>>>>>>>>>>");
  }

  fn simulate(&self) {
    self.pilot();
  }
}
